design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+data that needs to be loaded into memory begins.  The pointer is to the
+beginning of the last commit rather than the end, to make sure that the pointer
+is always contained within the DiskQueue.  This provides extra sanity checking
+on the validity of the DiskQueue's contents at recovery, at the cost of
+potentially reading 10MB more than what would be required.
+
+## Testing
+
+Correctness bugs in spilling would manifest as data corruption, which is well covered by simulation.
+The only special testing code added was to enable changing `log_spill` in `ConfigureTest`.
+This covers switching between spilling methods in the presence of faults.
+
+An `ASSERT` was added to simulation that verifies that commits read from the
+DiskQueue on recovery are only the commits which have not been spilled.
+
+The rest of the testing is to take a physical cluster and try the extremes that
+can only happen at scale:
+
+* Verify that recovery times are not impacted when a large amount of data is spilled
+* Verify that long running tests hit a steady state of memory usage (and thus there are likely no leaks).
+* Plot how quickly (MB/s) a remote datacenter can catch up in old vs new spilling strategy
+* See what happens when there's 1 tlog and more than 100 storage servers.
+  * Verify that peek requests get limited
+	* See if tlog commits can get starved by excessive peeking
+
+[fdbsummit-technical-overview]: https://www.youtube.com/watch?v=EMwhsGsxfPU
+
+# TLog Spill-By-Reference Operational Guide
+
+## Notable Behavior Changes
+
+TL;DR: Spilling involves less IOPS and is faster.  Peeking involves more IOPS and is slower.  Popping involves >0 IOPS.
+
+### Spilling
+
+The most notable effect of the spilling changes is that the Disk Queue files
+will now grow to potentially terrabytes in size.
+
+ 1. Spilling will occur in larger batches, which will result in a more
+sawtooth-like `BytesInput - BytesDurable` value.  I'm not aware that this will have any meaningful impact.
+
+ * Disk queue files will grow when spilling is happening
+   * Alerting based on DQ file size is no longer appropriate
+
+As a curious aside, throughput decreases as spilled volume increases, which
+quite possibly worked as accidental backpressure.  As a feature, this no longer
+exists, but means write-heavy workloads can drown storage servers faster than
+before.
+
+### Peeking
+
+Peeking has seen tremendous changes.  Its involves more IO operations and memory usage.
+
+The expected implication of this are:
+
+1.  A peek of spilled data will involve a burst of IO operations.
+
+		Theoretically, this burst can drown out queued write operations to disk,
+		thus and slowing down TLog commits.  This hasn't been observed in testing.
+
+		Low IOPS devices, such as HDD or network attached storage, would struggle
+		more here than locally attached SSD.
+
+2.  Generating a peek response of 150KB could require reading 100MB of data, and allocating buffers to hold that 100MB.
```

Comment:
Is this because you have to read the entire DiskQueue page to find the mutations for the tag that you care about in the DiskQueue?